### PR TITLE
FSF-3221 Integration Tests

### PR DESCRIPTION
**Pull request by andrelds11**

### CODE REVIEW

**Review by marhcouto — CHANGES_REQUESTED** [review]
Sorry for taking so long. Overall ok but tests need further documentation. Moreover, I don't think these were very thought through, we can discuss this tomorrow further

File: master/include/comm/communicator.hpp [hunk 1]
```diff
@@ -105,18 +107,17 @@ inline Communicator::Communicator() {
...
-    break_pressure *= HYDRAULIC_LINE_PRECISION; // convert back adding decimal part
-    _systemData->sensors._hydraulic_line_pressure = break_pressure;
-  } else if (buf[0] == RIGHT_WHEEL) {
+    _systemData->sensors._hydraulic_line_pressure = (buf[2] << 8) | buf[1];
+  }
+  else if (buf[0] == RIGHT_WHEEL) {
     double right_wheel_rpm = (buf[4] << 24) | (buf[3] << 16) | (buf[2] << 8) | buf[1];
     right_wheel_rpm *= WHEEL_PRECISION; // convert back adding decimal part
-    _systemData->sensors._rl_wheel_rpm = right_wheel_rpm;
+    _systemData->sensors._right_wheel_rpm = right_wheel_rpm;
```
[the remark this answers is not in the document]
Comment (andrelds11) [reply master/include/comm/communicator.hpp:120]: consistency

**Review by marhcouto — CHANGES_REQUESTED** [review]
We should talk about most of these comments. Main topics:
- the unit tests are giving me headaches
- a bit better documentation on the integration tests
- the errors mentioned
- not using function (to_ready) every time you can

File: master/test/test_integration/test_integration.cpp [hunk 2]
```diff
@@ -0,0 +1,505 @@
...
+    uint8_t hydraulic_msg[] = {HYDRAULIC_LINE, 0xf8, 0x00};
+    communicator.c1Callback(hydraulic_msg);
+
+    // Iterate a few times to go to check wd
+    Metro time{1};
+    while (!time.check()) {
+        as_state.calculateState();
+    }
+    
+    sd.digitalData.watchdog_state = false;
```
[the remark this answers is not in the document]
Comment (andrelds11) [reply master/test/test_integration/test_integration.cpp:96]: Watchdog needs to go low in the initial checkup sequence

File: master/test/test_integration/test_integration.cpp [hunk 3]
```diff
@@ -0,0 +1,505 @@
...
+        sd.failureDetection.inversorAliveTimestamp.reset();
+        sd.failureDetection.pcAliveTimestamp.reset();
+        sd.failureDetection.steerAliveTimestamp.reset();
+        sd.digitalData.watchdogTimestamp.reset();
+    }
+}
+
+/**
+ * @brief Test function to validate AS_OFF to AS_Ready transition 
+*/
```
[the remark this answers is not in the document]
Comment (andrelds11) [reply master/test/test_integration/test_integration.cpp:55]: Off to ready success transition with no other conditions or restraints

File: master/test/test_integration/test_integration.cpp [hunk 4]
```diff
@@ -0,0 +1,505 @@
...
+    }
+
+    TEST_ASSERT_EQUAL(false, went_ready);
+    TEST_ASSERT_EQUAL(State::AS_OFF, as_state.state);
+}
+
+/**
+ * @brief Test function to validate AS_READY doesn't revert to AS_Ready 
+*/
+void test_off_to_ready_wayback() {
```
[the remark this answers is not in the document]
Comment (andrelds11) [reply master/test/test_integration/test_integration.cpp:120]: It doesn't. The test is to validate it doesn't happen anymore. Changed the name slightly to make it clearer

File: master/test/test_integration/test_integration.cpp [hunk 5]
```diff
@@ -0,0 +1,505 @@
...
+    while (!time.check()){
+        as_state.calculateState();
+        sd.failureDetection.inversorAliveTimestamp.reset();
+        sd.failureDetection.pcAliveTimestamp.reset();
+        sd.failureDetection.steerAliveTimestamp.reset();
+        sd.digitalData.watchdogTimestamp.reset();
+    }
+
+    uint8_t msg[8] = {0x02, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00};
+    communicator.resStateCallback(msg);
```
[the remark this answers is not in the document]
Comment (andrelds11) [reply master/test/test_integration/test_integration.cpp:195]: Communicator is always used when possible for double validation (redundancy), maybe just with the exception of timestamps. Digital Data is always changed directly to do evaluate it with no breadboard inputs

File: master/test/test_integration/test_integration.cpp [hunk 6]
```diff
@@ -0,0 +1,505 @@
...
+    // brake pressure has a threshold to be updated, otherwise emergency
+    Metro time3{RELEASE_EBS_TIMEOUT_MS / 2};
+    while (!time3.check()){
+        sd.failureDetection.inversorAliveTimestamp.reset();
+        sd.failureDetection.pcAliveTimestamp.reset();
+        sd.failureDetection.steerAliveTimestamp.reset();
+        sd.digitalData.watchdogTimestamp.reset();
+        as_state.calculateState();
+    }
+    TEST_ASSERT_EQUAL(State::AS_DRIVING, as_state.state); // still within threshold, okay
```
[the remark this answers is not in the document]
Comment (andrelds11) [reply master/test/test_integration/test_integration.cpp:221]: Unit tests for the checkupmanager specific transitions are in the checkupmanager tests already

File: master/test/test_integration/test_integration.cpp [hunk 7]
```diff
@@ -0,0 +1,505 @@
...
+}
+
+/**
+ * @brief Test function to validate flow diagram conditions to go AS_DRIVING
+*/
+void test_flow_driving() {
+    reset();
+    TEST_ASSERT_EQUAL(State::AS_OFF, as_state.state);
+    uint8_t hydraulic_msg[] = {HYDRAULIC_LINE, 0xf8, 0x00};
+    communicator.c1Callback(hydraulic_msg);
```
[the remark this answers is not in the document]
Comment (andrelds11) [reply master/test/test_integration/test_integration.cpp:359]: Included in the initial checkup. Necessary for it to be ready. Will add a comment on that

File: master/test/test_integration/test_integration.cpp [hunk 8]
```diff
@@ -0,0 +1,505 @@
...
+    TEST_ASSERT_EQUAL(State::AS_OFF, as_state.state);
+}
+
+/**
+ * @brief Test function to validate flow diagram conditions to go AS_DRIVING
+*/
+void test_flow_driving() {
+    reset();
+    TEST_ASSERT_EQUAL(State::AS_OFF, as_state.state);
+    uint8_t hydraulic_msg[] = {HYDRAULIC_LINE, 0xf8, 0x00};
```
[the remark this answers is not in the document]
Comment (andrelds11) [reply master/test/test_integration/test_integration.cpp:358]: Should I use decimals? What is the idea?

**Review by marhcouto — APPROVED** [review]

**Comment by marhcouto** [comment]
I approved but the steering angle from steering's ID is wrong, it is 0x295D and it is an extended ID. Consult the PR in ROS-CAN if you need to check how it is read

**Comment by andrelds11** [comment]
> I approved but the steering angle from steering's ID is wrong, it is 0x295D and it is an extended ID. Consult the PR in ROS-CAN if you need to check how it is read

Sorry, I didn't notice this. I changed the id, from what I saw in the other rep, it is normally read so I just changed the id, can you confirm this?

**Pull request merged**